### PR TITLE
Markduckworth/or queries pr 9

### PR DESCRIPTION
**Pull request by MarkDuckworth**
- In expansion
- Public API w/o deprecation

### CODE REVIEW

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 1]
```diff
@@ -773,6 +773,466 @@ function genericQueryEngineTest(
...
+    verifyResult(result1, [doc1, doc4, doc5]);
+
+    const query2 = query('coll', filter('a', 'in', [2, 3]), orderBy('a'));
+
+    const result2 = await expectFunction(() =>
+      runQuery(query2, lastLimboFreeSnapshot)
+    );
+    verifyResult(result2, [doc6, doc3]);
+  });
+
```
Comment (ehsannas) [inline packages/firestore/test/unit/local/query_engine.test.ts:1235]: Thanks for adding all the additional tests.

IIUC `durable` means the test is using memory persistence.

so, what we're testing here are:
- queries using full collection scan (using memory persistence)
- queries using client-side indexing (using IndexedDB)

But, what we really need to have is:
- queries using full collection scan (using memory persistence)
- queries using full collection scan (using IndexedDB)
- queries using client-side indexing (using IndexedDB)

For Android, this is the case.
- Tests in QueryEngineTestCase.java are run twice, once for memory persistence (full collection scan + memory persistence) and once with SQLite persistence (full collection scan + sqlite).
- Tests in SQLiteQueryEngineTest.java are run once (CSI + sqlite)

**Review by ehsannas — COMMENTED** [review]
Looks good! Almost there

File: common/api-review/firestore-lite.api.md [hunk 2]
```diff
@@ -230,21 +230,28 @@ export class Query<T = DocumentData> {
...
     readonly type: 'or' | 'and';
 }
 
 // @public
 export abstract class QueryConstraint {
     abstract readonly type: QueryConstraintType;
+    /**
+     * Takes the provided {@link Query} and returns a copy of the {@link Query} with this
+     * {@link AppliableConstraint} applied.
+     */
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply common/api-review/firestore-lite.api.md:250]: Pretty strange, this was unexpected behavior in the autogenerated report from the api-report tool. I was able to work around it. Thanks for catching!

File: common/api-review/firestore.api.md [hunk 3]
```diff
@@ -359,21 +359,28 @@ export class Query<T = DocumentData> {
...
     readonly type: 'or' | 'and';
 }
 
 // @public
 export abstract class QueryConstraint {
     abstract readonly type: QueryConstraintType;
+    /**
+     * Takes the provided {@link Query} and returns a copy of the {@link Query} with this
+     * {@link AppliableConstraint} applied.
+     */
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply common/api-review/firestore.api.md:379]: Same fix.

File: packages/firestore/src/lite-api/query.ts [hunk 4]
```diff
@@ -84,67 +84,102 @@ export type QueryConstraintType =
...
+   */
+  abstract _apply<T>(query: Query<T>): Query<T>;
+}
 
 /**
  * A `QueryConstraint` is used to narrow the set of documents returned by a
  * Firestore query. `QueryConstraint`s are created by invoking {@link where},
- * {@link orderBy}, {@link (startAt:1)}, {@link (startAfter:1)}, {@link
- * (endBefore:1)}, {@link (endAt:1)}, {@link limit}, {@link limitToLast}, {@link or} or {@link and} and
+ * {@link orderBy}, {@link startAt:}, {@link startAfter}, {@link
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/lite-api/query.ts:104]: fixed

File: packages/firestore/src/lite-api/query.ts [hunk 5]
```diff
@@ -296,13 +331,40 @@ export class QueryCompositeFilterConstraint extends QueryFilterConstraint {
   }
 }
 
+/**
+ * `QueryNonFilterConstraint` is a helper union type that represents all
+ * QueryConstraints which do not inherit from QueryFieldFilterConstraint. These
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/lite-api/query.ts]: PTAL, I believe the new comment is more succinct and more accurate.

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 6]
```diff
@@ -773,6 +773,466 @@ function genericQueryEngineTest(
...
+        filter('a', 'in', [0, 1, 2, 4]),
+        filter('a', 'not-in', [1, 2])
+      )
+    );
+    const result2 = await expectFunction(() =>
+      runQuery(query2, lastLimboFreeSnapshot)
+    );
+    verifyResult(result2, []);
+
+    // a IN [] || a NOT-IN [0,2] should union them (similar to: a NOT-IN [0,2]).
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/test/unit/local/query_engine.test.ts]: Fixed. TY.

File: packages/firestore/test/unit/local/query_engine.test.ts (hunk 1 above)
Comment (MarkDuckworth) [reply packages/firestore/test/unit/local/query_engine.test.ts]: I refactored query_engine.test.ts so the tests execute in three configurations: 
1. durable with client side indexing
2. durable without client side indexing
3. non-durable without client side indexing

PTAL. Sorry, it's a little hard to read after refactoring and prettier.

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 7]
```diff
@@ -281,498 +290,713 @@ function genericQueryEngineTest(
...
+        LimitType.First
+      );
+      const results = await expectOptimizedCollectionQuery(() =>
+        runQuery(q, SnapshotVersion.min())
+      );
+
+      verifyResult(results, [doc1, doc2, doc4]);
+    });
+
+    it('or query with in and not-in using index', async () => {
```
Comment (ehsannas) [inline packages/firestore/test/unit/local/query_engine.test.ts:897]: `or query with in and not-in` test and `or query with in and not-in using index` test can be merged into one test (using `if (configureCsi)`.

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 8]
```diff
@@ -281,498 +290,713 @@ function genericQueryEngineTest(
...
+        'coll',
+        orFilter(filter('a', '==', 2), filter('b', 'not-in', [2, 3]))
+      );
+      const result2 = await expectOptimizedCollectionQuery(() =>
+        runQuery(query2, MISSING_LAST_LIMBO_FREE_SNAPSHOT)
+      );
+      verifyResult(result2, [doc1, doc2]);
+    });
+
+    it('query with array membership using index', async () => {
```
Comment (ehsannas) [inline packages/firestore/test/unit/local/query_engine.test.ts:949]: `query with array membership` test and `query with array membership using index` test can be merged into one test (using `if (configureCsi)`.

File: packages/firestore/test/unit/local/query_engine.test.ts (hunk 7 above)
Comment (MarkDuckworth) [reply packages/firestore/test/unit/local/query_engine.test.ts]: Good catch. This is updated.

File: packages/firestore/test/unit/local/query_engine.test.ts (hunk 8 above)
Comment (MarkDuckworth) [reply packages/firestore/test/unit/local/query_engine.test.ts]: Likewise, this is updeated.

**Review by ehsannas — APPROVED** [review]

**Pull request merged**